tutorials/graph-theory/connected-components.md

```diff
@@ -0,0 +1,126 @@
+---
+title: 'Connected-Components'
+description: 'Author: @adityabisht02'
+hide_table_of_contents: true
+keywords:
+  - leetcode
+  - tutorial
+  - prefix sum
+---
+
+<TutorialAuthors names="Aditya Bisht"/>
+
+## Overview
+In a graph, sometimes all nodes might not be connected with each other. Let's take an example of a graph with 5 nodes-
+
+
+![image](https://user-images.githubusercontent.com/89146189/195919108-2e06dbbe-717a-4f86-9ba1-90cbdf5c119c.png)
+
+Here you can see that 1-2-3 are connected and 4-5 are connected.
+Here , 1-2-3 is called one connected component.
+
+## How to determine if a connected component is present ?
+1. Start a BFS or DFS from any node of the graph.
+2. If you are able to visit all the other nodes from that single node that means the all nodes in the graph are connected.
+3. If you are unable to reach a particular node or a group of nodes, then that group of nodes might be a different connected component.
+
+## How to find connected components?
+1. Let's an adjacency list based graph is given.
+2. Traverse the adjacency list and for every node do a BFS or DFS on the node and mark all the nodes you visit in this way as visited.
+3. While traversing the list do not do BFS on the nodes which have already been visited.
+4. Hence everytime you do a BFS/DFS you are traversing a connected component so you can count the number of times you do a BFS/DFS.
+
+Let's understand with the help of an example:
+
+### Example : [0547 - Number Of Provinces](https://leetcode.com/problems/number-of-provinces/)
+
+```
+There are n cities. Some of them are connected, while some are not. If city a is connected directly with city b, and city b is connected directly with city c, then city a is connected indirectly with city c.
+
+A province is a group of directly or indirectly connected cities and no other cities outside of the group.
+
+You are given an n x n matrix isConnected where isConnected[i][j] = 1 if the ith city and the jth city are directly connected, and isConnected[i][j] = 0 otherwise.
+
+Return the total number of provinces.
+```
+
+Here, the given graph is in the form of an adjacency list. When we perform bfs/dfs starting from one particular node,
+we get one connected component. To make sure we get all connected components we need to traverse the adjacency list and make sure all nodes have been visited.
+
+<Tabs>
+<TabItem value="Java" label="Java">
+<SolutionAuthor name="@adityabisht02"/>
+
+```Java
+class Solution {
```

Comment:
format the code following [K&R Coding style](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)